### PR TITLE
Issue 296: Bookie supports ephemeral port

### PR DESCRIPTION
**Pull request by zhaijack**
Descriptions of the changes in this PR:

- add a flag to allow/disable using ephemeral ports
- change the initialization sequence to support ephmeral port
- added two tests on verifying this behavior

### CODE REVIEW

**Comment by zhaijack** [comment]
ported the change for @sijie

**Comment by eolivelli** [comment]
@zhaijack @sijie 
In DL tests do you need non-loopback addresses ? Another approach can be to allow port=0 (and automatic choose) on loopback interfaces (which in turn is to be enabled explicitly).

Another approach, only for tests, is to allow "local transport" (enableLocalTransport=true) and disabile "network transport" (disableServerSocketBind=true), this is one of the reasons for which I have introduced such features

**Comment by sijie** [comment]
@eolivelli 

we use loopback addresses. I don't see any reason to combine ephemeral  port with loopback address together. they have different purposes though.

ephemeral port is also different from local transport. using ephemeral port, we are exercising all the network transportation, which is important for all use cases.

**Comment by eolivelli** [comment]
@sijie I understand your point, I was only suggesting another approach, the local-transport is not good in your case.

What about not adding the new configuration option and allowing the use of ephemeral port on loopback address ? loopback is normally used only in tests
my concern is only to not add a dangerous option to be used in production.

**Comment by eolivelli** [comment]
@sijie  sorry I just re-read your comment.
OK for the new option, we are going to use non-loopback + ephemeral for tests

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java [hunk 1]
```diff
@@ -219,7 +225,12 @@ protected void initChannel(SocketChannel ch) throws Exception {
             });
 
             // Bind and start to accept incoming connections
-            bootstrap.bind(address.getAddress(), address.getPort()).sync();
+            Channel listen = bootstrap.bind(address.getAddress(), address.getPort()).sync().channel();
+            if (listen.localAddress() instanceof InetSocketAddress) {
+                if (conf.getBookiePort() == 0) {
+                    conf.setBookiePort(((InetSocketAddress) listen.localAddress()).getPort());
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java]: @jiazhai @sijie is there any other case in which we "change" the provided configuration object ?
wouldn't it be clearer to add a public getBookieSocketAddress() ?
if we do not have any other case it should be better to treat ServerConfiguration as immutable once the server is started

Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java]: @eolivelli currently bookie and netty server are separated, there is not cross referencing in between. The setBookiePort is propagated the real port to bookie for it to advertise its network location. you can't easily do getBookieSocketAddress() without any major refactor between bookie and netty server.

Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java]: OK for me

**Pull request closed without merging**